jina/types/arrays/document.py

```diff
@@ -299,6 +302,85 @@ def overriden_key(proto):
         else:
             self._pb_body.sort(*args, **kwargs)
 
+    def match(
+        self,
+        darray: 'DocumentArray',
+        metric: str = 'cosine',
+        limit: Optional[int] = inf,
+        is_distance: bool = False,
+    ) -> None:
+        """Compute embedding based nearest neighbour in `another` for each Document in `self`,
+        and store results in `matches`.
+
+        Note:
+            - If metric is 'cosine' it uses the cosine **distance**.
+            - If metric is 'euclidean' it uses the euclidean **distance**.
+            - If metric is 'euclidean_squared' it uses the euclidean **distance** squared.
+
+        :param darray: the other DocumentArray to match against
+        :param metric: the distance metric
+        :param limit: the maximum number of matches, when not given
+                      all Documents in `another` are considered as matches
+        :param is_distance: Boolean flag informing if `metric` values want to be considered as distances or scores.
+        """
+
+        def invert_if_score(
+            X: 'np.ndarray', metric: str, is_distance: bool
+        ) -> 'np.ndarray':
+            """Invert values to scores if `is_distance=False` according to the metric that is passed.
+
+            :param X: input np.ndarray
+            :param metric: string defining distance function ['cosine', 'euclidean', lidean_squared'].
+            :param is_distance: Boolean flag that describes if values in X need to be erted to similarities.
+            :return: A np.ndarray with distances if `is_distance=True`, scores if distance=False`.
+            """
+            if metric == 'cosine':
+                if is_distance:
+                    X = X
+                else:
+                    X = 1 - X
+            if metric == 'euclidean_squared' or metric == 'euclidean':
+                if is_distance:
+                    X = X
+                else:
+                    X = 1 / X
```

Comment:
we need to avoid exceptions with `division by zero`
```suggestion
                    X =  (1 / X + 1)
```